### ISSUE TITLE
dotnet build doesn't reliably build on m1 mac

### ISSUE THREAD

**Issue by eatonphil** [tigerbeetle/tigerbeetle#762]
See the log here: https://gist.github.com/eatonphil/df78a29152a9ce3f2f4af35401dcea0e. Even after clearing the global and local cache it doesn't build very reliably. Sometimes after running `dotnet build` twice it works. Sometimes after running `dotnet build` six times.

**Comment by eatonphil** [comment]
I'm pretty sure part of it is cross-compilation. With this diff, `dotnet build` works reliably.

```
diff --git a/build.zig b/build.zig
index 8bab8357..954d1194 100644
--- a/build.zig
+++ b/build.zig
@@ -692,26 +692,32 @@ fn dotnet_client(
     bindings.setMainPkgPath("src");
     const bindings_step = bindings.run();
 
+    const arm_mac = builtin.cpu.arch == .aarch64 and builtin.os.tag == .macos;
+
     inline for (platforms) |platform| {
         const cross_target = CrossTarget.parse(.{ .arch_os_abi = platform[0], .cpu_features = "baseline" }) catch unreachable;
 
-        const lib = b.addSharedLibrary("tb_client", "src/clients/c/tb_client.zig", .unversioned);
-        lib.setMainPkgPath("src");
-        lib.setOutputDir("src/clients/dotnet/TigerBeetle/runtimes/" ++ platform[1] ++ "/native");
-        lib.setTarget(cross_target);
-        lib.setBuildMode(mode);
-        lib.linkLibC();
+        // TODO(Phil): ARM macs having trouble cross-compiling.
+        const can_build = (arm_mac and cross_target.cpu_arch.?.isAARCH64() and cross_target.isDarwin()) or !arm_mac;
+        if (can_build) {
+            const lib = b.addSharedLibrary("tb_client", "src/clients/c/tb_client.zig", .unversioned);
+            lib.setMainPkgPath("src");
+            lib.setOutputDir("src/clients/dotnet/TigerBeetle/runtimes/" ++ platform[1] ++ "/native");
+            lib.setTarget(cross_target);
+            lib.setBuildMode(mode);
+            lib.linkLibC();
 
-        if (cross_target.os_tag.? == .windows) {
-            lib.linkSystemLibrary("ws2_32");
-            lib.linkSystemLibrary("advapi32");
-        }
+            if (cross_target.os_tag.? == .windows) {
+                lib.linkSystemLibrary("ws2_32");
+                lib.linkSystemLibrary("advapi32");
+            }
 
-        lib.addOptions("vsr_options", options);
-        link_tracer_backend(lib, tracer_backend, cross_target);
+            lib.addOptions("vsr_options", options);
+            link_tracer_backend(lib, tracer_backend, cross_target);
 
-        lib.step.dependOn(&bindings_step.step);
-        build_step.dependOn(&lib.step);
+            lib.step.dependOn(&bindings_step.step);
+            build_step.dependOn(&lib.step);
+        }
     }
 }
 
diff --git a/src/clients/dotnet/TigerBeetle/TigerBeetle.csproj b/src/clients/dotnet/TigerBeetle/TigerBeetle.csproj
index 92701d19..6b56cef6 100644
--- a/src/clients/dotnet/TigerBeetle/TigerBeetle.csproj
+++ b/src/clients/dotnet/TigerBeetle/TigerBeetle.csproj
@@ -16,7 +16,7 @@
     <Exec Command=".\scripts\build.bat dotnet_client -Drelease-safe" WorkingDirectory="$(ProjectDir)\..\..\..\.." />
   </Target>
   <Target Name="BuildZig" BeforeTargets="CoreCompile" Condition="'$(OS)'!='Windows'">
-    <Exec Command="zig/zig build dotnet_client -Drelease-safe" WorkingDirectory="$(ProjectDir)/../../../.." />
+    <Exec Command="./scripts/build.sh dotnet_client -Drelease-safe" WorkingDirectory="$(ProjectDir)/../../../.." />
   </Target>
   <ItemGroup>
     <Content Include="runtimes\**\*.so">

```

**Comment by eatonphil** [comment]
Seems related to https://github.com/ziglang/zig/issues/9711.

**Comment by batiati** [comment]
> Seems related to [ziglang/zig#9711](https://github.com/ziglang/zig/issues/9711).

It is related, and we used to have a workaround for that:
https://github.com/tigerbeetledb/tigerbeetle-java/blob/195ca1f948a329564f71faea53907863a1fd5725/src/zig/build.zig#L37-L43

We removed that code because it was flawed, overwriting the cache dir does not affect each build step, as effect all artifacts were built using the same last directory set. Here is a [better solution for isolating cache folders](https://github.com/tigerbeetledb/tigerbeetle/blob/0a57f90186b0784bc1d1ebff59f16089aedacf41/build.zig#L970-L1009). It potentially solves this issue and that one building the Java client on Windows.

But still, there's something about the way `dotnet` executes the [csproj `<Exec>` tag](https://github.com/tigerbeetledb/tigerbeetle/blob/0a57f90186b0784bc1d1ebff59f16089aedacf41/src/clients/dotnet/TigerBeetle/TigerBeetle.csproj#L15-L20) that randomly fails in another point.
Needs more investigation.
